pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,671 @@
+		}
+
+		samples := 0
+		firstSample := true
+		prevTS := int64(-1)
+
+		it := ch.Iterator(nil)
+		for it.Err() == nil && it.Next() {
+			samples++
+			ts, _ := it.At()
+
+			if firstSample {
+				firstSample = false
+				if ts != cm.MinTime {
+					// TODO: Error?
+					level.Warn(c.logger).Log("ref", cm.Ref, "msg", "timestamp of the first sample doesn't match chunk MinTime",
+						"sampleTimestamp", formatTimestamp(ts), "chunkMinTime", formatTimestamp(cm.MinTime))
+				}
+			} else if ts <= prevTS {
+				// TODO: Error?
+				level.Warn(c.logger).Log("ref", cm.Ref, "msg", "found sample with timestamp not strictly higher than previous timestamp",
+					"previous", formatTimestamp(prevTS), "sampleTimestamp", formatTimestamp(ts))
+			}
+
+			prevTS = ts
+		}
+		if e := it.Err(); e != nil {
+			return errors.Wrapf(err, "failed to failed to iterate over samples of chunk %d", cm.Ref)
+		}
+		if samples == 0 {
+			// TODO: Error?
+			level.Warn(c.logger).Log("ref", cm.Ref, "msg", "no samples found in the chunk")
+		} else if prevTS != cm.MaxTime {
+			// TODO: Error?
+			level.Warn(c.logger).Log("ref", cm.Ref, "msg", "timestamp of the last sample doesn't match chunk MaxTime",
+				"sampleTimestamp", formatTimestamp(prevTS), "chunkMaxTime", formatTimestamp(cm.MaxTime))
+		}
+	}
+
+	return nil
+}
+
+func (c *MultitenantCompactor) sanitizeMeta(tenantID, blockID string, meta *metadata.Meta) error {
+	if meta.Thanos.Labels == nil {
+		meta.Thanos.Labels = map[string]string{}
+	}
+
+	metaULID := meta.ULID
+	if metaULID.String() != blockID {
+		level.Warn(c.logger).Log("msg", "updating meta.json block ID",
+			"old_value", metaULID.String(), "new_value", blockID, "user", tenantID)
+		var err error
+		meta.ULID, err = ulid.Parse(blockID)
+		if err != nil {
+			return errors.Wrapf(err, "couldn't parse block ID %q", blockID)
+		}
+	}
+
+	metaTenantID := meta.Thanos.Labels[mimir_tsdb.TenantIDExternalLabel]
+	if metaTenantID != tenantID {
+		level.Warn(c.logger).Log("msg", "updating meta.json tenant label", "block_id", blockID,
+			"old_value", metaTenantID, "new_value", tenantID, "user", tenantID)
+		meta.Thanos.Labels[mimir_tsdb.TenantIDExternalLabel] = tenantID
+	}
+
+	var rejLbls []string
+	for l, v := range meta.Thanos.Labels {
+		switch l {
+		case mimir_tsdb.TenantIDExternalLabel, mimir_tsdb.CompactorShardIDExternalLabel:
+		case mimir_tsdb.IngesterIDExternalLabel, mimir_tsdb.DeprecatedShardIDExternalLabel:
+			level.Debug(c.logger).Log("msg", "removing unused external label from meta.json",
+				"block_id", blockID, "user", tenantID, "label", l, "value", v)
+			delete(meta.Thanos.Labels, l)
+		default:
+			rejLbls = append(rejLbls, l)
+		}
+	}
+
+	if len(rejLbls) > 0 {
+		level.Warn(c.logger).Log("msg", "rejecting unsupported external label(s) in meta.json",
+			"block_id", blockID, "user", tenantID, "labels", strings.Join(rejLbls, ","))
+		return errBadRequest{message: fmt.Sprintf("unsupported external label(s): %s", strings.Join(rejLbls, ","))}
```

Comment:
Why use "errBadRequest" wrapper? Any error from this method is user-problem.

Reply:
@pstibrany right now it will only return a user-caused error, it's true. But what about in the future, when a developer might add an internal error and doesn't think of the assumption that the function only returns used-caused errors? I think making this assumption would be very fragile.